Functions/Quantification/xASL_quant_HadamardDecoding.m

```diff
@@ -107,98 +125,141 @@
+indexControl = indexControl(:)';
 
-nRepetitions = nEncodedVolumes / (xQ.TimeEncodedMatrixSize * xQ.NumberEchoTimes); % Calculating no. of acquisition repeats
+decodedControl4D = imEncoded(:,:,:, indexControl);
+xQ.EchoTime_Control4D = xQ.EchoTime(indexControl);
+xQ.InitialPLD_Control4D = xQ.Initial_PLD(indexControl);
+xQ.LabelingDuration_Control4D = xQ.LabelingDuration(indexControl);
 
-%% 2. Reorder multi-TE data
+%% 4. Reorder image matrix
 % At this point the data is organized like this (in terms of ASL4D.nii volumes):
 % PLD1/TE1,PLD1/TE2,PLD1/TE3,PLD1/TE4...PLD2/TE1,PLD2/TE2,PLD2/TE3... (TEs in first dimension, PLDs after)
 %
 % And for decoding we want
 % TE1/PLD1,TE1/PLD2,TE1/PLD3,TE1/PLD4...TE2/PLD1,TE2/PLD2,TE2/PLD3,TE2/PLD4 (PLDs in the first dimension, TEs after)
-if isfield(xQ,'NumberEchoTimes') && (xQ.NumberEchoTimes > 1)
+% Then below, we apply this "transformation"/"decoding" vector to the image matrix and to the parameters TE/PLD/LD
+if isfield(xQ, 'NumberEchoTimes') && (xQ.NumberEchoTimes > 1)
 	% This is the original order
-	vectorOldOrder = 1:nEncodedVolumes;
+	vectorReorderEncoded = 1:nEncodedVolumes;
 	
 	% Shape to a matrix with TEs first and all the rest later
-	vectorOldOrder = reshape(vectorOldOrder, xQ.NumberEchoTimes, nEncodedVolumes/xQ.NumberEchoTimes);
+	vectorReorderEncoded = reshape(vectorReorderEncoded, xQ.NumberEchoTimes, nEncodedVolumes/xQ.NumberEchoTimes);
 	
 	% Flip the two dimensions and make a row vector again
-	vectorOldOrder = reshape(vectorOldOrder', 1, nEncodedVolumes);
+	vectorReorderEncoded = reshape(vectorReorderEncoded', 1, nEncodedVolumes);
 
 	% Reorder the data
-	imEncoded = imEncoded(:,:,:,vectorOldOrder);
+	imEncodedReordered = imEncoded(:,:,:,vectorReorderEncoded);
 end
 
 % Sometimes HAD8 is PLD1(rep1), PLD1(rep2), PLD2(rep1), PLD2(rep2). We
 % want: PLD1(rep1), PLD2(rep1)... PLD1(rep2),PLD2(rep2)
-% UniquePLDs = unique(xQ.Initial_PLD);
-% 
-% if (length(UniquePLDs) == nEncodedVolumes) && (UniquePLDs(2)-UniquePLDs(1))< 250
-%     This is the original order
-% 	vectorOldOrder = 1:nEncodedVolumes;
-%     Shape
-%     vectorOrder = [ vectorOldOrder(1:2:end) vectorOldOrder(2:2:end)];
-%     Reorder the data
-% 	imEncoded = imEncoded(:,:,:,vectorOrder);
-% end
-
-%% 3. Decode the Hadamard data
-imDecoded = zeros(size(imEncoded,1), size(imEncoded,2), size(imEncoded,3), nDecodedVolumes * nRepetitions);    
-idx=0;
-for iRepetition = 1:nRepetitions
-    for iTE = 1:xQ.NumberEchoTimes
-        for iTI = 1:nDecodedTI
-            
-            indexPositive = find(xQ.TimeEncodedMatrix(iTI,:)==1);
-            indexNegative = find(xQ.TimeEncodedMatrix(iTI,:)==-1);
-            imDecoded(:,:,:,((iTE-1)*nDecodedTI+iTI)+(iRepetition-1)*nDecodedVolumes) = sum(imEncoded(:,:,:,(indexPositive+idx)),4) - sum(imEncoded(:,:,:,(indexNegative+idx)),4);
-            
-        end
-        idx = idx+xQ.TimeEncodedMatrixSize;
-    end
-end
 
+%% 5. Decode the Hadamard data
```

Comment:
"Decode the Hadamard data" -> "Decode the image volumes"

Reply:
Agree. Will do.